Modules/_testcapi/unicode.c

```diff
@@ -220,109 +220,61 @@ unicode_copycharacters(PyObject *self, PyObject *args)
     return Py_BuildValue("(Nn)", to_copy, copied);
 }
 
-/* Test PyUnicode_ToLower() */
 static PyObject *
-unicode_tolower(PyObject *self, PyObject *arg)
+unicode_case_operation(PyObject *str, int (*function)(Py_UCS4, Py_UCS4 *, int), const char *name)
 {
-    if (PyUnicode_GET_LENGTH(arg) != 1) {
-        PyErr_SetString(PyExc_ValueError, "unicode_tolower only accepts 1-character strings");
+    if (PyUnicode_GET_LENGTH(str) != 1) {
+        PyErr_Format(PyExc_ValueError, "%s only accepts 1-character strings", name);
         return NULL;
     }
 
-    Py_UCS4 c = PyUnicode_READ_CHAR(arg, 0);
+    Py_UCS4 c = PyUnicode_READ_CHAR(str, 0);
 
-    Py_UCS4 lower[3];
-    int chars = PyUnicode_ToLower(c, lower, Py_ARRAY_LENGTH(lower));
-    assert(chars >= 1);
+    Py_UCS4 buf[3];
+    int chars = function(c, buf, Py_ARRAY_LENGTH(buf));
+    if (chars <= 0) {
+        PyErr_BadInternalCall();
+        return NULL;
+    }
 
     PyUnicodeWriter *writer = PyUnicodeWriter_Create(1);
     if (writer == NULL) {
         return NULL;
     }
-    if (PyUnicodeWriter_WriteUCS4(writer, lower, chars) < 0) {
+    if (PyUnicodeWriter_WriteUCS4(writer, buf, chars) < 0) {
         PyUnicodeWriter_Discard(writer);
         return NULL;
     }
     return PyUnicodeWriter_Finish(writer);
```

Comment:
Maybe use `PyUnicode_FromKindAndData(PyUnicode_4BYTE_KIND, buf, chars);` instead of PyUnicodeWriter to save a few lines.